File: _archived/notifications-nodejs/src/integrations/webhooks/normalizers/twilio.normalizer.ts
import { NormalizedEventType } from "../../../models/notification-event.model";

export interface TwilioStatusPayload {
  MessageSid?: string;
  SmsSid?: string;
  SmsStatus?: string;
  MessageStatus?: string;
  To?: string;
  From?: string;
  ErrorCode?: string;
  ErrorMessage?: string;
  AccountSid?: string;
  [key: string]: string | undefined;
}

export interface NormalizedProviderEvent {
  rawEventType: string;
  normalizedEventType: NormalizedEventType;
  providerMessageId: string | null;
  eventTimestamp: Date;
  recipientPhone: string | null;
  errorCode: string | null;
  errorMessage: string | null;
  metadata: Record<string, unknown>;
}

const TWILIO_STATUS_MAP: Record<string, NormalizedEventType> = {
  queued: "queued",
  accepted: "accepted",
  sending: "queued",
  sent: "sent",
  receiving: "queued",
  received: "delivered",
  delivered: "delivered",
  undelivered: "undeliverable",
  failed: "failed",
  canceled: "failed",
  scheduled: "queued",
  read: "opened",
};

export function normalizeTwilioEvent(raw: TwilioStatusPayload): NormalizedProviderEvent {
  const rawStatus = (raw.MessageStatus || raw.SmsStatus || "unknown").toLowerCase();
  const normalizedEventType: NormalizedEventType = TWILIO_STATUS_MAP[rawStatus] ?? "failed";

  const messageSid = raw.MessageSid || raw.SmsSid || null;

  const { MessageSid, SmsSid, SmsStatus, MessageStatus, ...rest } = raw;

  return {
    rawEventType: rawStatus,
    normalizedEventType,
    providerMessageId: messageSid ?? null,
    eventTimestamp: new Date(),
    recipientPhone: raw.To ?? null,
    errorCode: raw.ErrorCode ?? null,
    errorMessage: raw.ErrorMessage ?? null,
    metadata: rest,
  };
}

File: apps/services/notifications/Notifications.Infrastructure/Webhooks/Normalizers/TwilioNormalizer.cs
namespace Notifications.Infrastructure.Webhooks.Normalizers;

public class NormalizedTwilioEvent
{
    public string RawEventType { get; set; } = string.Empty;
    public string NormalizedEventType { get; set; } = string.Empty;
    public string? ProviderMessageId { get; set; }
    public DateTime EventTimestamp { get; set; }
    public string? RecipientPhone { get; set; }
    public string? ErrorCode { get; set; }
    public string? ErrorMessage { get; set; }
    public Dictionary<string, object> Metadata { get; set; } = new();
}

public static class TwilioNormalizer
{
    private static readonly Dictionary<string, string> StatusMap = new()
    {
        ["queued"] = "queued",
        ["accepted"] = "accepted",
        ["sending"] = "queued",
        ["sent"] = "sent",
        ["receiving"] = "queued",
        ["received"] = "delivered",
        ["delivered"] = "delivered",
        ["undelivered"] = "undeliverable",
        ["failed"] = "failed",
        ["canceled"] = "failed",
        ["scheduled"] = "queued",
        ["read"] = "opened"
    };

    public static NormalizedTwilioEvent Normalize(Dictionary<string, string> formParams)
    {
        var rawStatus = (formParams.GetValueOrDefault("MessageStatus") ?? formParams.GetValueOrDefault("SmsStatus") ?? "unknown").ToLowerInvariant();
        var normalizedEventType = StatusMap.GetValueOrDefault(rawStatus, "failed");
        var messageSid = formParams.GetValueOrDefault("MessageSid") ?? formParams.GetValueOrDefault("SmsSid");

        return new NormalizedTwilioEvent
        {
            RawEventType = rawStatus,
            NormalizedEventType = normalizedEventType,
            ProviderMessageId = messageSid,
            EventTimestamp = DateTime.UtcNow,
            RecipientPhone = formParams.GetValueOrDefault("To"),
            ErrorCode = formParams.GetValueOrDefault("ErrorCode"),
            ErrorMessage = formParams.GetValueOrDefault("ErrorMessage"),
            Metadata = new Dictionary<string, object>()
        };
    }
}
